hpy/devel/src/runtime/ctx_type.c

```diff
@@ -352,7 +353,8 @@ create_slot_defs(HPyType_Spec *hpyspec, HPy_ssize_t base_member_offset,
     hpyslot_count++;        // Py_tp_methods
     hpyslot_count++;        // Py_tp_members
     hpyslot_count++;        // Py_tp_getset
-    hpyslot_count++;        // Py_tp_dealloc
+    if (has_tp_traverse(hpyspec) || has_tp_destroy(hpyspec))
+        hpyslot_count++;        // Py_tp_dealloc
```

Comment:
I fear that this style (which was introduced by me, so I'm the culprit :)) might quickly become too hard to read/maintain, especially because the logic needs to be kept in sync with the lines 435-438 below.
Maybe introducing a helper `needs_py_tp_dealloc()` will result in better code?

Reply:
Good point, done!